incubator/hnc/pkg/controllers/object_controller_new.go

```diff
@@ -38,25 +59,306 @@ type ObjectReconcilerNew struct {
+	// This object is the source.
+	if !isPropagated(inst) {
+		log.Info("This object is the source", "namespace", inst.GetNamespace())
+		r.syncSource(ctx, log, inst)
+		// Early exit.
+		return unchanging, nil
+	}
+
+	// This object is a propagated copy.
+	log.Info("This object is a propagated copy", "namespace", inst.GetNamespace())
+	return r.syncPropagated(ctx, log, inst)
+}
+
+// syncPropagated will delete obsolete copy or overwrite itself with the source.
+// Or do nothing if it remains the same as the source object.
+func (r *ObjectReconcilerNew) syncPropagated(ctx context.Context, log logr.Logger, inst *unstructured.Unstructured) (action, *unstructured.Unstructured) {
+	so := r.Forest.Get(inst.GetNamespace()).GetSource(r.GVK, inst.GetName())
+
+	// Delete the obsolete copy and early exit if there's no source in the ancestors.
+	if so == nil {
+		return deleting, nil
+	}
+
+	// If the copy is different from the source, overwrite it.
+	if !reflect.DeepEqual(object.Canonical(inst), object.Canonical(so)) {
+		//ns := inst.GetNamespace()
+		//inst = object.Canonical(so)
+		//inst.SetNamespace(ns)
+		metadata.SetLabel(inst, api.LabelInheritedFrom, so.GetNamespace())
+		return overwriting, so
+	}
+
+	return unchanging, nil
+}
+
+// syncSource syncs the copy in the forest with the current source object. If there's a change,
+// enqueue all the descendants to propagate the new source.
+func (r *ObjectReconcilerNew) syncSource(ctx context.Context, log logr.Logger, src *unstructured.Unstructured) {
+	// syncPropagated doesn't check isExcluded because the propagation won't happen if it's excluded.
+	if r.isExcluded(log, src) {
+		return
+	}
+	sns := src.GetNamespace()
+	n := src.GetName()
+	origCopy := r.Forest.Get(sns).GetOriginalObject(r.GVK, n)
+
+	// Early exit if the source object exists and remains unchanged.
+	if origCopy != nil && reflect.DeepEqual(object.Canonical(src), object.Canonical(origCopy)) {
+		log.Info("Unchanged Source")
+		return
+	}
+
+	// Update or create a copy of the source object in the forest
+	r.Forest.Get(sns).SetOriginalObject(src.DeepCopy())
+
+	// Enqueue all the descendant copies
+	r.enqueueDescendants(ctx, log, src)
 }
 
 // SyncNamespace can be called manually by the HierarchyReconciler when the hierarchy changes.
+// It enqueues all the current objects in the namespace and local copies of the original objects
+// in the ancestors.
 func (r *ObjectReconcilerNew) SyncNamespace(ctx context.Context, log logr.Logger, ns string) error {
-	log.Info("SyncNamespace placeholder")
+	log = log.WithValues("gvk", r.GVK)
+	ul := &unstructured.UnstructuredList{}
+	ul.SetGroupVersionKind(r.GVK)
+	if err := r.List(ctx, ul, client.InNamespace(ns)); err != nil {
+		log.Error(err, "Couldn't list objects")
+		return err
+	}
+
+	// Enqueue all the current objects in the namespace because some of them may have been deleted.
+	for _, inst := range ul.Items {
+		if r.isExcluded(log, &inst) {
+			continue
+		}
+		co := object.Canonical(&inst)
```

Comment:
Probably worth noting that we don't actually _need_ the entire canonical object here, just its metadata. But this is the easiest way to do it.